stdlib/public/Concurrency/AsyncStream.swift

```diff
@@ -338,6 +338,10 @@ public struct AsyncStream<Element> {
   ) {
     let storage: _AsyncStreamCriticalStorage<Optional<() async -> Element?>>
       = .create(produce)
+
+    let cancelStorage: _AsyncStreamCriticalStorage<(() -> Void)?>?
+      = if let onCancel { .create(onCancel) } else { nil }
```

Comment:
i'm not sure introducing another lock is the ideal approach, but it seems at least serviceable. it wasn't immediately obvious to me if `_AsyncStreamCriticalStorage` supports a fully generic payload, or if it's supposed to contain only a closure (or closure-sized generic parameter?) per [this comment](https://github.com/swiftlang/swift/blob/f40d3b4c2e5b7ee9d161e06d9fdc31543336a3d8/stdlib/public/Concurrency/AsyncStreamBuffer.swift#L541). also i tried changing to use `Mutex` directly, but could not get the tests to successfully build, so unsure if that approach would be an option.

Reply:
I'd try to make a `_Storage` struct and make it have onCancel as well as the other state in there; so we don't have to have separate locks like this